### DISCUSSION TITLE
Progress reporting order guarantee question

### DISCUSSION THREAD

**Discussion by rjacobw** [microsoft/vs-streamjsonrpc#1077] [Q&A] [answered]
How strong is the [guarantee ](https://github.com/microsoft/vs-streamjsonrpc/blob/main/doc/progresssupport.md#timing-considerations) that progress is reported to a client in the order in which the server reports it? Is there a limit on how quickly progress can be reported before this guarantee is no longer supported? I seem to be able to get progress reported out of order if reported very quickly over an async named pipe.

I have a very simple interface:
```
public interface IApplication : IDisposable
{
    Task PrintText(string text, IProgress<int> progress);
}
```

and very simple server method:

```
public Task PrintText(string text, IProgress<int> progress)
{
    for (var i = 0; i <= 100; i += 10)
    {
        // Insert a delay and progress is reported in the correct order
        progress.Report(i);
    }

    Console.WriteLine(text);
    return Task.CompletedTask;
}
```

and a very simple client method (running in a separate process):

```
var progress = new Progress<int>(i =>
{
    Console.WriteLine($"Progress: {i}");
});

await application.PrintText("This is coming from the client", progress);
```
When I run the client and server without any type of progress reporting delay in the server code, I can debug the progress action and see that progress is reported out of order - similar to the following:

![Screenshot 2024-09-17 143951](https://github.com/user-attachments/assets/18650b81-6fa4-4200-9ebf-8ec4c9e212da)

This is obviously a contrived example where progress is reported very quickly (much more so than is typically reported in real scenarios), but I'm just curious if there's a known limit on how quickly progress can/should be reported (or if I'm doing something incorrectly that is causing the progress to be reported out of order).

**Comment by AArnott** [comment, marked as answer]
Great question. 

Please find the answer in https://github.com/microsoft/vs-streamjsonrpc/blob/main/doc/progresssupport.md.

In short, yes the order is guaranteed, insofar as RPC is involved. In your contrived example, order is being lost within your own client code because the delegate you provided that prints the callbacks runs on the threadpool, which doesn't guarantee order.
